### ISSUE TITLE
Add possibility to use break and continue in f# loops. And do while loop

### ISSUE THREAD

**Issue by baronfel** [baronfel/fsharp-lang-suggestions-final#381]
# Add possibility to use break and continue in f# loops. And do while loop [6923281] #

**Submitted by Vladimir on 1/5/2015 12:00:00 AM**  
**12 votes on UserVoice prior to migration**  

It very often looks better than recursion, and it would be wonderful to have this possibility

**[Original UserVoice Submission](https://fslang.uservoice.com/forums/245727-f-language/suggestions/6923281)**
[Archived Uservoice Comments](../tree/master/archive/suggestion-6923281-add-possibility-to-use-break-and-continue-in-f-l.md#comments)

**Comment by AlOdeychuk** [comment]
The break and continue keywords are long-awaited additions to the F# language due to pragmatic reasons (migration of codebases from Java, C++ or another imperative language, friendliness of the language syntax to programmers with imperative programming background, i.e. the majority of new and existing language users). The break and continue statements should have two forms: labeled and unlabeled like in Java, the TIOBE TOP1 language. Labels attached to loops have to be declared in the Java style: a label name followed by a colon. Please see details at https://docs.oracle.com/javase/tutorial/java/nutsandbolts/branch.html

**Comment by DemiMarie** [comment]
I agree, but for a different reason: performance (break and continue may very well be faster).

**Comment by smoothdeveloper** [comment]
@DemiMarie I think this would need substanciation if we could provide samples highlighting the issue.

From discussions with seasoned F# developers, it seems recursive function or while loop with a boolean condition would have pretty similar cost.

The feature was mentionned by @migueldeicaza : https://www.infoq.com/articles/virtual-panel-dotnet-future

> For F#, my request is simple: I am a man that is too attached to his old ways of writing loops.   I want my loops to include support for break and continue. I know this is heresy, but that is what I desire the most :-)

Well, going on the road of performance, and not introducing many arbitrary constructs, what about `goto` and labels? `break` and `continue` are limited to the closest loop which is not always what you want.

So far I haven't missed those constructs, and I think they are not obvious to introduce gracefully in expression based F#.

It would be nice to have a summary of downvoters motives.

**Comment by BentTranberg** [comment]
I am against this. Use of "break" (except in "case"), "continue", "goto" and "return" in C# can make it hard to understand the meaning, especially in large methods where use of these keywords can be hard to spot.

You should be able to quickly make quite a few assumptions just by taking a glance at the structure of the code. These assumptions easily break when any of these mechanisms are allowed. For that reason alone they should all be banned in F#.

Although simple moderate sensible use of "break" (and "return") is very popular and de facto standard pattern in C#, and definitely makes code shorter and more readable, there is no need to have this in F#, since equally simple functional constructs exist. Translating from C# to F# is very simple in these cases.

C# source can be rewritten to not use them, and then easily be ported to F#. If the C# code is hard to rewrite, then it just shows the code had low quality to begin with, and it justifies why it should not be so in F#.

I have translated tens of thousands of lines of C# to F#. The biggest stumbling blocks by far is with "continue", heavy use of "break" in odd places in loops, and not least "goto". Even if these were available in F#, it would be even more important for me to refactor them away as quickly as possible in order to take advantage of the language. In hard cases, that refactoring is easier done in C# anyway.

Although F# is a multi-paradigm language, it is also a functional first language. These mechanisms are impure even in the imperative domain. They would influence even the functional constructs in a negative way, because they make it harder to reason about the code in total.

**Comment by isaksky** [comment]
I think this, as well as adding normal imperative "return" would be a great change to F#. Right now in F# there are a few alternatives:

1. Increase nesting (if - else), pattern matching
   - This option is problematic, because it does not scale well as you add more conditions.
2. Wrap results and compose with functions like `Option.map`
   - This option is not satisfactory, because it is more verbose, adds allocations, and is less direct than just returning early
3. Computation expressions
   - [Not efficient](http://www.quanttec.com/fparsec/users-guide/where-is-the-monad.html)
   - Would require an ad-hoc computation expression per type you want to compose ( see `asyncMaybe`, for example. That is not elegant, it is ad-hoc and crazy.)
     - Even this wouldn't work for multiple different types (e.g., `ParseResult` and `HttpRequestResult`)

Some people say `let rec` and small functions is an alternative, but in that case you are at best shuffling the problem around. It is like telling someone callback hell is fine, because you can just split the logic up into more functions.

**Comment by greatim** [comment]
The following code may help:

```
module Range =

    let inline (|Range|) (x: Range) = (x.Start.Value, x.End.Value)
    
    let inline range a b = Range(Index a, Index b)

open Range
type LoopBuilder() =

    member inline __.For(s: seq<_>, f) =
        use s = s.GetEnumerator()
        let mutable result = ValueNone
                       
        while s.MoveNext() && 
            (   result <- f s.Current
                result.IsNone 
            ) do ()
        result

    member inline __.For(Range(x, y), f) =
        let mutable result = ValueNone
        let mutable i = x
        while i < y do 
            result <- f i
            i <- if result.IsNone then 
                    i + 1
                  else 
                    y
        result


    member inline __.For(s: IReadOnlyList<'t>, f) =
        let len = s.Count
        let mutable result = ValueNone
        let mutable i = 0

        while i < len do
            result <- f s.[i]
            i <- if result.IsNone then 
                    i + 1                    
                  else
                    len
                    
        result
                

    member inline __.Return x = ValueSome x

    member inline __.Zero() = ValueNone

    member inline __.Yield x = x

    member inline __.Combine(x, f) =
        match x with
        | ValueSome x -> x
        | ValueNone -> f()

    member inline __.Delay f = f

    member inline __.Run f = f()

 let loop = LoopBuilder() 
```
**usage:**
```
   // used `yield`,  it return the true type value
    let foo2 (arr: int[]) = loop {
        for x in arr do
            for y in arr do 
                if x * y > 10000 then 
                    return x + y
        yield 0
    }
   // without  `yileld`, it return a `voption` value
    let foo3 (arr: _ List) = loop {
        for x in arr do
            if x > 10000 then
                return x
    }


   **// It's hard to say, it don't support ` for x in 0 .. 10000 do`, when you write that, it call the `seq` overload.
   // can somebody tell me how?**

    let foo4 (xs: Range) = loop {
        for x in xs do
            if x > 10000 then
                return x
    }
```

This code looks ugly. But I compare the many times, it make the most clean result:
```
public static FSharpValueOption<int> foo3(List<int> arr)
{
	int count = ((IReadOnlyCollection<int>)arr).Count;
	FSharpValueOption<int> result = FSharpValueOption<int>.ValueNone;
	for (int num = 0; num < count; num = ((result.Tag != 0) ? count : (num + 1)))
	{
		result = f@38-3(((IReadOnlyList<int>)arr)[num]);
	}
	return result;
}
internal static FSharpValueOption<int> f@38-3(int _arg1)
{
	if (_arg1 > 10000)
	{
		return FSharpValueOption<int>.NewValueSome(_arg1);
	}
	return FSharpValueOption<int>.ValueNone;
}
```

I do not agree that the calculation expression must be inefficient.
However, it would be better if F# added the `break return` syntax.

**Comment by yatli** [comment]
@greatim my poorman's break is always `async{ ... }` without actual async code :p

**Comment by greatim** [comment]
> @greatim my poorman's break is always `async{ ... }` without actual async code :p
Oh no!
It's hard to use, and packed with several layers of function.

**Comment by isral** [comment]
@yatli 
Please give sample code

**Comment by yatli** [comment]
simply:
```F#
async {
  while true do
    if foo then return
}
```

**Comment by greatim** [comment]
> simply:
> 
> ```fsharp
> async {
>   while true do
>     if foo then return
> }
> ```

It doesn't compile

**Comment by yatli** [comment]
Uh oh. But why?
I _thought_ I understand CE. 😅😅

**Comment by greatim** [comment]
> Uh oh. But why?
> I _thought_ I understand CE. 😅😅

hmmm, I can see why now. I fixed it for this:
```
      let foo x =
         async {
           while true do
             if !x then return ()
         }
```

but the loop won't be breaked.

**Comment by yatli** [comment]
@greatim yeah that's my confusion -- why doesn't it stop at return? What does it even mean to return multiple times from async? That the async CE can be run multiple times, statefully?

**Comment by greatim** [comment]
> What does it even mean to return multiple times from async?

I don't know what it does.
 But last time I told my friend that "using `async` can break the loop".
 He did it, and hit my face on the next day.

**Comment by BentTranberg** [comment]
I think as a general rule we should keep posts in these open repos (fsharp/fslang-suggestion, dotnet/fsharp, etc) focused on the issue, so that we won't have to waddle through unrelated stuff repeatedly. In my experience one tends to read or skim through the whole thread from start to end several times. I suggest you sort out side issues like this in one of the channels in F# Slack, unless you feel it's actually relevant to the issue.

**Comment by daveyostcom** [comment]
> I am against this. Use of "break" (except in "case"), "continue", "goto" and "return" in C# can make it hard to understand the meaning, especially in large methods where use of these keywords can be hard to spot.

IDEs should have a separate category for coloring these keywords.  I would use red.

**Comment by jl0pd** [comment]
@nkosi23 I don't want to sound toxic, but I don't understand how lack of break can improve readability/maintainability especially in such case

```fsharp
    // https://github.com/dotnet/fsharp/blob/c281bd08b2d6bcbf60f5ccaa925f17de8ef714f1/src/FSharp.Core/array.fs#L490
    let inline contains value (array: 'T array) =
        checkNonNull "array" array
        let mutable state = false
        let mutable i = 0

        while not state && i < array.Length do
            state <- value = array.[i]
            i <- i + 1

        state
```

```fsharp
    let inline contains value (array: 'T array) =
        checkNonNull "array" array

        for el in array do
             if el = value then
                   break true
        else 
            false
```

Break is a natural way to stop loops and lack of it makes F# worse

**Comment by jwosty** [comment]
Some sort of ``iterWhile`` sounds promising.

Throwing my hat in the ring here - I rarely run into this issue, but I am currently running into this today while translating some [SDF raymarching](https://iquilezles.org/articles/raymarchingdf/) shader code to F# (so that I can evaluate this SDF on both the CPU and the GPU).

Here's the code (HLSL):

```hlsl
RaymarchResult shortestDistanceToSurface(float3 rayOrigin, float3 rayDirection, float start, float end) {
    float depth = start;
    float minDist = end;
    RaymarchResult rr = (RaymarchResult)0;
    for (rr.i = 0; rr.i < MAX_MARCHING_STEPS; rr.i++) {
        SdfResult sr = sceneSdf(rayOrigin + depth * rayDirection);
        if (sr.dist < minDist) {
            minDist = sr.dist;
            rr.nearDepth = depth;
        }
        if (sr.dist < EPSILON) {
            sr.dist = minDist;
            rr.depth = depth;
            rr.sdfResult = sr;
            return rr;
        }
        depth += sr.dist;
        if (depth >= end) {
            sr.dist = minDist;
            rr.depth = end;
            rr.sdfResult = sr;
            return rr;
        }
    }
    rr.depth = end;
    rr.sdfResult.dist = minDist;
    rr.sdfResult.c = float3(0,0,0);
    return rr;
}
```

Not the best code, for sure, but it's proving to be quite annoying to translate due to the early returns/breaks and mutable variables. Just thought it may be useful to provide another real-life sample.

For anyone who's interested in more details: The main awkwardness is because I need to make sure both the F# version and the HLSL version do the same thing, so every further step I take away from the HLSL version equates to more mental calculations. Granted, it's a pattern smell to have these two implementations side-by-side, and the ideal would either be to unit test both implementations, and that's a rabbit hole I'm sure I will be forced to go down sooner or later... but for now, this is what I'm burdening myself with. Not the end of the world, but probably an `iterWhile` would make this a little less awkward.

**Comment by MarcoPeli** [comment]
Please implement return + continue + break. The idea of a language shouldn't be to force developers go functional/OO/imperative/etc. The idea should be to allow developers choose freely how they want things done. When I need to process a big file (lot of GBs) I can't afford putting millions of "None" inside a collection. Because it has to be fast, I need to just go to the next item (continue) or to break immediately without a cumbersome while + flag. Sometimes I need to break entirely because a condition was met, but there is no return keyword. I need to test the data and process or skip what shouldn't be processed. Without return/break/continue it's just too hard to do those things in the context of BIG data processing. In the big data and LLMs era, not having return/continue/break is killing the language.

Please stop trying to force developers into one style or another. Give developers the freedom they deserve. I'm looking into F# instead of C# because writing quick and small functions in the functional style is very intuitive. But in C# I have to write "public static async Task< sometype> myFun (more types...) " what's just too annoying. We developers are lazy, otherwise we wouldn't automate things. In F# I can do "let myFun = ..." that's a lot like python, the number 1 lang of the world. Learn from it, please. Writing python is FUN (but slow execution). Anyway, without return/break/continue the F# language is just unable to handle big data processing in an ergonomic way. I hope you see.

**Comment by nkosi23** [comment]
@MarcoPeli For the cases you mention, the idiomatic way is to use functions such as List.choose, no need to add empty items to your list. F# is very capable for big data processing and this is one of the many areas where it shines (eg. Type providers).

Your issues seem to be mainly due to the fact that you want to use imperative programming in an ML language. C# is a better language for that, i think you should use F# only if you value the benefits that a .NET ML-like language can provide. And if this the case, there definitely is a learning curve to move from an imperative mindset to a functional mindset. I have been there. 

You won't be able to express your problems the same way, but this is precisely the point IMHO, since this provides you guarantees that you cannot have with imperative programming. Your code is also easier to reason about and maintain.

Now regarding your broader point about forcing a particular programming style, i understand where you are coming but i think that instead of seeing these restrictions as limitations, we can understand that they are most of the time constraints providing correctness and/or interpretation guarantees. Knowing that F# is expression based and that early returns are not possible provides a number of guarantees and makes code easier to reason about. Many of these guarantees would be severely weakened by supporting the keywords you mention.

Beyond this, I'd also note that F# has long been opionionated. For example the virtual keyword is not supported to avoid promoting inheritance while adding support for it would be relatively low effort. IMHO, this is great as the language is very flexible but is subtly skewed toward functional patterns. In my view, the language has found the perfect balance between being a multi-paradigm language, while remaining an expresssion-based functional language providing a range of guaranteed and promoting functional approaches.

Not dismissing what you are saying just offering a different perspective.

**Comment by BentTranberg** [comment]
Even though F# is described as being a multi-paradigm language, the idea of F# is clearly functional first, and next object programming rather than object oriented programming. It is a pragmatic solution to functional programming on the .NET platform.

Even in C# and other OOP languages, object oriented programming has been toned down during the last decades. Object programming - avoiding complex inheritance - is a better way. And now they're even adopting functional features.

From what I remember reading, they experimented with Haskell first in .NET and found it too frustrating. Basing F# on ML was better. This part of the story also tells us functional was the point.

No, I absolutely don't want an F# that invites neither imperative nor OOP more than strictly necessary, because I want efficient and safe coding in the team. That is _the reason_ I stick to F#. I also believe the compiler can do a much better job opmimizing source that can be reasoned about to a much higher degree, so let's not break that.

Lack of break and continue doesn't mean the code has to be slower. I once had a heated argument with a C/C++ programmer who argued fervently C/C++ was much better than Delphi because statements like x += 1 and x++ in C++ made it more efficient. There was no convincing him that optimizing compilers made this syntax difference completely irrelevant. Regardless of how F# vs C# code looks before and after they are compiled, it isn't these things that matter the most. We'll (the F# community) always have the option of adding more optimization to the compiler, and of the perhaps 0.01% of code that needs extreme optimization, we (ourselves) can compile it in a highly optimizing C++ compiler or in assembly, if we (the sufficiently wealthy) find that buying a faster computer can't fix it for a much lower cost and in less time.

**Comment by T-Gro** [comment]
I want to give a "probably not" on a language change, while approving specific library additions to accomplish the motivations listed here. `iterWhile` for `Array;List;Seq` and one for `Range` (exact API and module placement for ranges to be designed and discussed via an RFC. Tuples would feel natural for expressing those for me: `(1,99) |> Range.iterWhile(fun i ....)`).


Since this is a performance-oriented suggestion, an important criteria is to provide a version of `Array.iterWhile` that (after inlining) leads to IL which the JIT can recognize and eliminate bounds checks for.

**Comment by vzarytovskii** [comment]
Yeah, it's "probably not", it has been decided in 1st versions of the language. Additional library functions would be the way to go.

**Comment by smoothdeveloper** [comment]
Makes me think contributions to the "F# guidelines", that shows how to implement properly, equivalent of control flow structure some people seem to be missing, could also fit the job.

In my own experience, I've not felt the lack was ever a show stopper, and the clarity of expression based, and plain old simple control flow that F# has is worth it, however it may be limitating for specific edge cases, for which, there is work around.

One thing I'd consider still, is a do/loop construct, that would be idiomatic to F# design.

**Comment by MarcoPeli** [comment]
See how this new lang: https://www.moonbitlang.com/ is doing exactly as I said on previous comment: Being friendly for functional and for imperative styles, gives developers the freedom to choose the better style according to the task at hand.

I'll put a reminder on my calendar for 3 years from today, to continue this discussion. If this new lang gets more adoption/popularity than F# I have a point. If not, I'll agree you have a better point keeping the "purity" of the F# language.
